File: PYTHON/nest_digital/binary_search.py
def search(nums, target) -> int:
    start = 0
    end = len(nums)-1
    while start <= end:
        mid = int(start + (end - start)/2)
        if(nums[mid] == target):
            return mid
        if(nums[mid] < target):
            start = mid+1
        else:
            end = mid -1

    return -1
nums = [-1,0,3,5,9,12]
target = 9
print(search(nums,target))

File: JAVA/Searching/BinarySearch.java
public class BinarySearch {
    public static void main(String[] args) {
        int[] arr = {-1,0,2,4,13,19,24,35,47,66,89,97};
        int target = 66;
        int res = search(arr, target);
        System.out.println(res);
    }
    static int search(int[] arr, int target) {
        int start = 0;
        int end = arr.length-1;
        while(start <= end) {
            int mid = start + (end - start) / 2; // to avoid int size exceeding
            if(target > arr[mid]) {
                start = mid + 1;
            } else if(target < arr[mid]) {
                end = mid - 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
}
